sympy/angles/euler.py

```diff
@@ -0,0 +1,688 @@
+        r"""Returns Euler angles representing same rotation as the quaternion,
+        in the sequence given by `seq`. This implements the method described
+        in [1]_.
+
+        Parameters
+        ==========
+
+        q : Quaternion or list
+            Rotation quaternion elements.
+
+        seq : string of length 3
+            Represents the sequence of rotations.
+            For intrinsic rotations, seq must be all lowercase and its elements
+            must be from the set `{'x', 'y', 'z'}`
+            For extrinsic rotations, seq must be all uppercase and its elements
+            must be from the set `{'X', 'Y', 'Z'}`
+
+        angle_addition : bool
+            Default : True
+            When True, first and third angles are given as an addition and
+            subtraction of two simpler `atan2` expressions. When False, the
+            first and third angles are each given by a single more complicated
+            `atan2` expression. This equivalent is given by:
+
+            --math::
+
+                \operatorname{atan_2} (b,a) \pm \operatorname{atan_2} (d,c) =
+                \operatorname{atan_2} (bc\pm ad, ac\mp bd)
+
+        avoid_square_root : bool
+            Default : False
+            When True, the second angle is calculated with an expression based
+            on `acos`, which is slightly more complicated but avoids a square
+            root. When False, second angle is calculated with `atan2`, which
+            is simpler and can be better for numerical reasons (some
+            numerical implementations of `acos` have problems near zero).
+
+        Returns
+        =======
+
+        Euler
+            The Euler angles calculated from the quaternion
+
+        Examples
+        ========
+
+        >>> from sympy import Quaternion, Euler
+        >>> from sympy.abc import a, b, c, d
+        >>> q = Quaternion(a, b, c, d)
+        >>> euler = Euler.from_quaternion(q, seq='zyz')
+        >>> euler
+        Euler((-atan2(-b, c) + atan2(d, a)),  2*atan2(sqrt(b**2 + c**2), sqrt(a**2 + d**2)),  (atan2(-b, c) + atan2(d, a)), zyz)
+
+        >>> euler = Euler.from_quaternion([a, b, c, d], seq='zyz')
+        >>> euler
+        Euler((-atan2(-b, c) + atan2(d, a)),  2*atan2(sqrt(b**2 + c**2), sqrt(a**2 + d**2)),  (atan2(-b, c) + atan2(d, a)), zyz)
+
+        References
+        ==========
+
+        .. [1] https://doi.org/10.1371/journal.pone.0276302
+
+        """
+        if not isinstance(q, Quaternion):
+            if not iterable(q):
+                q = [q]
+            q = Quaternion(*q)
+
+        angles = q.to_euler(seq,
+                            angle_addition=angle_addition,
+                            avoid_square_root=avoid_square_root)
+        return Euler(*angles, seq=seq)
+
+    def add(self, other):
+        """Adds Euler angles elements element-wise, treating it as a Matrix.
+
+        This does not have much mathematical meaning, are you sure this is what
+        you want?
```

Comment:
This does seem a bit odd

Reply:
The sum of two rotation matrices isn't a rotation matrix, the same way that a sum of a rotation quaternion is not a quaternion.

I was not sure if it was a good idea, but I decided to implement the sum of Euler angles in the same vein: adding all of the elements, but not necessarily giving back something as meaningful as the other values.

I can totally remove all of the definitions of additions though.

Reply:
I think that multiplication can also be left undefined for the cases where it must need coversion to matrix or quaternion.

Reply:
I think that composition with other angles, although inefficient, could be a good addition to this class that would at least work well for numerical values or simple cases.

What if I separate it in a subfunction called `Euler.compose` and raise a `ValueError` inside of  `__mul__` when two Euler values are being multiplied, giving out a message that says `Multiplication undefined between two Euler instances. For rotation composition, use the 'compose' method.`?

Reply:
Why define these methods at all? Are they actually needed for something?

Reply:
You mean the addition/multiplication as matrices, or the composition?

Reply:
Maybe the least ambiguous implementation would be for `__mul__` to be implemented as a call to `compose` only when both elements are instances of `Euler`.

For other inputs of `__mul__`, and any input of  `__add__` or `__sub__`, the usual type error with a message saying that for treating the elements as a `Matrix`, the user should use `to_Matrix()`.

And `__pow__` could keep being the composition with itself as already implemented.

Reply:
I think if there's any ambiguity about what operations like `+` and `*` mean then it is better not to define them at all. If the operations are just going to convert to matrix or quaternion then it would be better for the user to use matrices or quaternions.

What is this class going to be used for? Don't just add methods for the sake of it.

It isn't clear to me that this class should be added at all.

Reply:
This implementation is a rather simple one for now, because I wanted to first see if there was any interest in such a class.

The idea with it is to further on implement some helpers for dealing with Euler angles. For example, one thing that I find particularly confusing but is used a LOT in research papers is to use Euler angles when deriving the Equations of Motion of a body, and then using Euler angles again to derive a control law for them.

With quaternions, this is easy, since, for an angular velocity $\pmb \omega$:
```math
\dot q = \frac{1}{2} q \odot \begin{bmatrix} 0 \\ \pmb \omega \end{bmatrix} 
```

With Euler angles, having an expression of their derivaties is a LOT more complicated, because for three angles $\alpha$, $\beta$ and $\gamma$, what we have is:
```math
\begin{align}
\dot R &= \pmb A \begin{bmatrix} \dot \alpha \\ \dot \beta \\ \dot \gamma \end{bmatrix} \\
\dot R &= R \hat{\pmb \omega}
\end{align}
```

Where this matrix $\pmb A$ is always a complicated expression that depends on the sequence used.

One thing I have planned is to implement a helper for calculating these matrices for example, to help me navigate these research papers a bit better.

Since I'm probably not the only one having to deal with this, then I imagined I'd first see if such an Euler angle helper class could be of any interest in Sympy.